Simplenote/src/main/java/com/automattic/simplenote/models/Tag.java

```diff
@@ -79,54 +79,59 @@ public void setIndex(Integer tagIndex) {
 
     public void renameTo(String tagOld, String tagNew, int index, Bucket<Note> notesBucket) throws BucketObjectNameInvalid {
         // When old tag ID is equal to new tag hash, tag is being renamed to lexical variation.
-        boolean isOldIdEqualToNewHash = getSimperiumKey().equals(TagUtils.hashTag(tagNew));
+        boolean isRenamingToLexicalTag = getSimperiumKey().equals(TagUtils.hashTag(tagNew));
         //noinspection unchecked
         Bucket<Tag> tagsBucket = (Bucket<Tag>) getBucket();
-        // Get all notes with old tag to update.
-        ObjectCursor<Note> notes = findNotes(notesBucket, tagOld);
-
-        while (notes.moveToNext()) {
-            Note note = notes.getObject();
-            List<String> tagsNew = new ArrayList<>();
-            List<String> tagsHash = new ArrayList<>();
-
-            // Create lists of note's tags excluding old tag.
-            for (String tag : note.getTags()) {
-                if (!tag.equals(tagOld)) {
-                    tagsNew.add(tag);
-                    tagsHash.add(TagUtils.hashTag(tag));
-                }
+
+        if (isRenamingToLexicalTag) {
+            if (!getName().equals(tagNew)) {
+                setName(tagNew);
+                save();
+            }
+        } else {
+            // Create new tag if canonical tag doesn't already exist.
+            if (!TagUtils.hasCanonicalOfLexical(tagsBucket, tagNew)) {
+                TagUtils.createTag(tagsBucket, tagNew, index);
             }
 
-            // Add lexical tag to note.  Update this tag's name and save it.
-            if (isOldIdEqualToNewHash) {
-                tagsNew.add(tagNew);
-
-                if (!getName().equals(tagNew)) {
-                    setName(tagNew);
-                    save();
-                }
-            // Add new canonical tag to note and create new tag.  Delete this tag.
-            } else {
-                // Add new tag if note doesn't already have same hashed tag.
-                if (!tagsHash.contains(TagUtils.hashTag(tagNew))) {
-                    tagsNew.add(TagUtils.getCanonicalFromLexical(tagsBucket, tagNew));
-                }
-
-                // Create new tag if canonical tag doesn't already exist.
-                if (!TagUtils.hasCanonicalOfLexical(tagsBucket, tagNew)) {
-                    TagUtils.createTag(tagsBucket, tagNew, index);
-                }
-
-                delete();
+            delete();
```

Comment:
let's keep the delete operation at the end of the sequence. we don't want to inadvertently trigger an unwanted sequence of updates on another device when it sees a tag disappear, before we've updated the notes referencing the tag